packages/core/data-transfer/src/directory/providers/source/index.ts

```diff
@@ -0,0 +1,208 @@
+import type { Readable } from 'stream';
+
+import path from 'path';
+import { PassThrough, pipeline } from 'stream';
+import fs from 'fs-extra';
+import { isEmpty, keyBy } from 'lodash/fp';
+import { chain } from 'stream-chain';
+import { parser } from 'stream-json/jsonl/Parser';
+import type { Struct } from '@strapi/types';
+
+import type { IAsset, IMetadata, ISourceProvider, ProviderType, IFile } from '../../../../types';
+
+import * as utils from '../../../utils';
+import { ProviderInitializationError, ProviderTransferError } from '../../../errors/providers';
+
+/**
+ * Constant for the metadata file path
+ */
+const METADATA_FILE_PATH = 'metadata.json';
+
+/**
+ * Provider options
+ */
+export interface ILocalDirectorySourceProviderOptions {
+  directory: {
+    path: string; // the directory to load files from
+  };
+}
+
+export const createLocalDirectorySourceProvider = (
+  options: ILocalDirectorySourceProviderOptions
+) => {
+  return new LocalDirectorySourceProvider(options);
+};
+
+class LocalDirectorySourceProvider implements ISourceProvider {
+  type: ProviderType = 'source';
+
+  name = 'source::local-directory';
+
+  options: ILocalDirectorySourceProviderOptions;
+
+  #metadata?: IMetadata;
+
+  constructor(options: ILocalDirectorySourceProviderOptions) {
+    this.options = options;
+  }
+
+  /**
+   * Pre flight checks to ensure the directory exists and can be read.
+   */
+  async bootstrap() {
+    const { path: directoryPath } = this.options.directory;
+
+    try {
+      // Check if the directory exists
+      const stat = await fs.stat(directoryPath);
+      if (!stat.isDirectory()) {
+        throw new Error(`Path is not a directory: ${directoryPath}`);
+      }
+      // Load the metadata file
+      await this.#loadMetadata();
+    } catch (e) {
+      throw new ProviderInitializationError(
+        `Could not initialize provider: ${(e as Error).message}`
+      );
+    }
+
+    if (!this.#metadata) {
+      throw new ProviderInitializationError('Could not load metadata from the directory.');
+    }
+  }
+
+  async #loadMetadata() {
+    const metadataPath = path.join(this.options.directory.path, METADATA_FILE_PATH);
+    this.#metadata = await this.#readJSONFile<IMetadata>(metadataPath);
+  }
+
+  async #loadAssetMetadata(filePath: string) {
+    return this.#readJSONFile<IFile>(filePath);
+  }
+
+  async getMetadata() {
+    if (!this.#metadata) {
+      await this.#loadMetadata();
+    }
+
+    return this.#metadata ?? null;
+  }
+
+  async getSchemas() {
+    const schemaCollection = await utils.stream.collect<Struct.Schema>(
```

Comment:
Kind of annoying we have to do this (read => convert to stream => convert to a collection) since we could basically do a simple
`concat(files.map(f => fs.read(f)).split('\n').map(l => JSON.parse(l))` (kinda)

But your impl keeps consistency with everything else, so let's stick with it